{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Basic Unix\n",
    "__1.__ `ls -al`  \n",
    "__2.__ There are 4 files in the directory (and 2 subdirectories). You can use `ls -l` to look inside the directory. This will show you which of the contents are files and which are directories. Don't forget to also include the `-a` option to show any hidden files:  \n",
    "   \n",
    "`ls -la Pfalciparum`  \n",
    "  \n",
    "__3.__ Malaria.fa is the largest file. You can add the `-h` option to the command above to make the size of the files more readable.  \n",
    "  \n",
    "__4.__ `cd Pfalciparum`  \n",
    "  \n",
    "__5.__ There is one file in the fasta directory (hint: it's a hidden file!).  \n",
    "  \n",
    "__6.__ `cp Pfalciparum.bed annotation`  \n",
    "  \n",
    "__7.__ `mv *.fa fasta`\n",
    "  \n",
    "__8.__ 4 files.  \n",
    "  \n",
    "__9.__ There are 6 GFFs in the Unix directory. To search from the Unix directory, you can either use `cd` to move up to the directory, or you can specify the path in the `find` command. This can either be the absolute path, which you can get from `pwd`, or you can use the relative path, like so:  \n",
    "   \n",
    "`find ../.. -name *.gff`\n",
    "  \n",
    "__10.__ There are 7 fasta files in the Unix directory. Note that fasta files normally end with `.fa` OR `.fasta`, so wee need to make sure we look for both of these, by adding a wildcard after `fa`:   \n",
    "   \n",
    "`find ../.. -name *.fa*`   "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "# Looking inside files\n",
    "__1.__ `head -n 500 Styphi.gff > Styphi.500.gff`  \n",
    "  \n",
    "__2.__ There are 6213 lines in the file. Use the `-l` option:  \n",
    "  \n",
    "`wc -l Pfalciparum.bed`  \n",
    "  \n",
    "__3.__ `sort -k 1 -k 2 -n Pfalciparum.bed`  \n",
    "  \n",
    "__4.__ Here is one way to do this: First use `awk` to get the first column of the file. Sort this and then use the `-c` option for `uniq` to count how many entries each chromosome has:\n",
    "\n",
    "`awk '{ print $1 }' Pfalciparum.bed | sort | uniq -c`\n",
    "\n",
    "The expected output would be:\n",
    "\n",
    "```\n",
    " 190 01\n",
    " 264 02\n",
    " 287 03\n",
    " 292 04\n",
    " 357 05\n",
    " 373 06\n",
    " 395 07\n",
    " 374 08\n",
    " 425 09\n",
    " 452 10\n",
    " 553 11\n",
    " 621 12\n",
    " 773 13\n",
    " 857 14\n",
    " ```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Searching inside files with `grep`\n",
    "\n",
    "__1.__ `grep \"^>\" exercises.fasta`  \n",
    "  \n",
    "__2.__ There are 1000 sequences. We can use -c to count the number of matches:  \n",
    "  \n",
    "`grep -c \"^>\" exercises.fasta`  \n",
    "  \n",
    "Or pipe into wc:\n",
    "  \n",
    "`grep \"^>\" exercises.fasta | wc -l`  \n",
    "    \n",
    "__3.__ Yes, three of them:  \n",
    "\n",
    "```\n",
    ">sequence27 spaces in the name\n",
    ">sequence52 another with spaces\n",
    ">sequence412 yet another with spaces\n",
    "```\n",
    "  \n",
    "One option is two greps piped together:  \n",
    "   \n",
    "`grep \"^>\" exercises.fasta | grep \" \"`  \n",
    "   \n",
    "Alternatively, in one regular expression  \n",
    "   \n",
    "`grep \"^>.* .*\" exercises.fasta`  \n",
    "  \n",
    "__4.__ `grep -v \"^>\" exercises.fasta`    \n",
    "   \n",
    "__5.__ Three. First extract the sequences, then search for n:  \n",
    "   \n",
    "`grep -v \"^>\" exercises.fasta | grep -c -i n`  \n",
    "   \n",
    "__6.__ Yes, one sequence. Try:   \n",
    "  \n",
    "`grep -v \"^>\" exercises.fasta | grep -i -v \"^[acgtn].*$\"`  \n",
    "  \n",
    "Alternatively, we can use the ^ to ask for matches NOT in the alphabet [acgtn]  \n",
    "  \n",
    "`grep -v \"^>\" exercises.fasta | grep -i \"[^ACGTN]\"`  \n",
    "  \n",
    "__7.__ 66 sequences. Try:  \n",
    "  \n",
    "`grep -v \"^>\" exercises.fasta | grep -c \"GC[AT]GC\"`  \n",
    "  \n",
    "__8.__ We found the total number of sequences earlier:  \n",
    "    \n",
    "`grep -c \"^>\" exercises.fasta`\n",
    "  \n",
    "... which outputs 1000  \n",
    "  \n",
    "This finds the number of unique names:  \n",
    "  \n",
    "`grep \"^>\" exercises.fasta | sort | uniq | wc -l`  \n",
    "  \n",
    "... which outputs 999.  \n",
    "  \n",
    "Therefore there is 1000 - 999 = 1 name repeated.  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# File processing with `awk`\n",
    "\n",
    "__1.__ Using:  \n",
    "  \n",
    "`awk -F\"\\t\" '{print $1}' exercises.bed | sort -u`  \n",
    "  \n",
    "Should give you:  \n",
    "```\n",
    "contig-1\n",
    "contig-3\n",
    "contig-4\n",
    "contig-5\n",
    "scaffold-2\n",
    "```\n",
    "  \n",
    "__2.__ There are 5 contigs. Use the command from the previous exercise and count the number of lines with `wc`:  \n",
    "   \n",
    "`awk -F\"\\t\" '{print $1}' exercises.bed | sort -u | wc -l`  \n",
    "  \n",
    "__3.__ There are 164 features on the positive strand. Try:  \n",
    "   \n",
    "`awk -F\"\\t\" '$6==\"+\"' exercises.bed | wc -l`  \n",
    "   \n",
    "__4.__ Ther are 124 features on the negative strand. Try:  \n",
    "   \n",
    "`awk -F\"\\t\" '$6==\"-\"' exercises.bed | wc -l`  \n",
    "   \n",
    "__5.__ There are 293 genes. Try:\n",
    "  \n",
    "`awk -F\"\\t\" '$4 ~ /gene/' exercises.bed | wc -l`  \n",
    "   \n",
    "__6.__ 5 genes have no strand assigned to them. Try:  \n",
    "  \n",
    "`awk -F\"\\t\" '$4 ~ /gene/ && $6 != \"-\" && $6 != \"+\"' exercises.bed | wc -l`\n",
    "  \n",
    "__7.__ Yes (6 of them are). First, the number of genes was found earlier:  \n",
    "  \n",
    "`awk -F\"\\t\" '$4 ~ /gene/' exercises.bed | wc -l`  \n",
    "  \n",
    "The number of unique names is:  \n",
    "  \n",
    "`awk -F\"\\t\" '$4 ~ /gene/ {print $4}' exercises.bed | sort -u | wc -l`  \n",
    "  \n",
    "Alternatively, the names can be found like this:  \n",
    "  \n",
    "`awk -F\"\\t\" '$4 ~ /gene/ {print $4}' exercises.bed  | sort | uniq -c | awk '$1>1'`  \n",
    "  \n",
    "__9.__ 18541. Try:  \n",
    "  \n",
    "`awk -F\"\\t\" '$4==\"repeat\" {score+=$5} END {print score}' exercises.bed`  \n",
    "  \n",
    "__10.__ There are 75 features in contig-1. Try:  \n",
    "  \n",
    "`awk -F\"\\t\" '$1 == \"contig-1\"' exercises.bed  | wc -l`  \n",
    "   \n",
    "__11.__ There are 12 repeats in contig-1. Try:  \n",
    "  \n",
    "`awk -F\"\\t\" '$1 == \"contig-1\" && $4 == \"repeat\"' exercises.bed  | wc -l`\n",
    "  \n",
    "__12.__ The mean score is 560.833. Try:   \n",
    "   \n",
    "`awk -F\"\\t\" '$1 == \"contig-1\" && $4 == \"repeat\" {score+=$5; count++} END{print score/count}' exercises.bed`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# BASH scripts\n",
    "__1.__ Here is an example of what this script could look like:\n",
    "```\n",
    "#!/usr/bin/env bash\n",
    "set -e\n",
    "\n",
    "# check that the correct number of options was given.\n",
    "# If not, then write a message explaining how to use the\n",
    "# script, and then exit.\n",
    "if [ $# -ne 1 ]\n",
    "then\n",
    "    echo \"usage: example_1.sh filename\"\n",
    "    echo\n",
    "    echo \"Prints the number of lines in the file\"\n",
    "    exit\n",
    "fi\n",
    "\n",
    "# Use sensibly named variable\n",
    "filename=$1\n",
    "\n",
    "# check if the input file exists\n",
    "if [ ! -f $filename ]\n",
    "then\n",
    "    echo \"File '$filename' not found! Cannot continue\"\n",
    "    exit\n",
    "fi\n",
    "\n",
    "\n",
    "# If still here, we can count the number of lines\n",
    "number_of_lines=$(wc -l $filename | awk '{print $1}')\n",
    "echo \"There are $number_of_lines lines in the file $filename\"\n",
    "```\n",
    "   \n",
    "__2.__ Here is an example of what this script could look like:\n",
    "```\n",
    "#!/usr/bin/env bash\n",
    "set -e\n",
    "for filename in ../scripts/loop_files/*; do ./exercise_1.sh $filename; done\n",
    "```\n",
    "\n",
    "__3.__ Here is an example of what this script could look like:\n",
    "```\n",
    "#!/usr/bin/env bash\n",
    "set -e\n",
    "\n",
    "# Check if the right number of options given.\n",
    "# If not, print the usage\n",
    "if [ $# -ne 1 ]\n",
    "    then\n",
    "        echo \"usage: example_3.sh in.gff\"\n",
    "        echo\n",
    "        echo \"Gathers some summary information from a gff file\"\n",
    "        exit\n",
    "fi\n",
    "\n",
    "# store the filename in a better named variable\n",
    "infile=$1\n",
    "\n",
    "\n",
    "# Stop if the input file does not exist\n",
    "if [ ! -f $infile ]\n",
    "then\n",
    "    echo \"File '$infile' not found! Cannot continue\"\n",
    "    exit 1\n",
    "fi\n",
    "\n",
    "echo \"Gathering data for $infile...\"\n",
    "\n",
    "\n",
    "# Gather various stats on the file...\n",
    "\n",
    "\n",
    "# Total number of lines/records in file\n",
    "total_records=$(wc -l $infile | awk '{print $1}')\n",
    "echo \"File has $total_records records in total\"\n",
    "\n",
    "\n",
    "# Get the sources from column 2.\n",
    "echo\n",
    "echo \"The sources in the file are:\"\n",
    "awk '{print $2}' $infile | sort -u\n",
    "\n",
    "\n",
    "# Count the sources\n",
    "echo\n",
    "echo \"Count of sources, sorted by most common\"\n",
    "awk '{print $2}' $infile | sort | uniq -c | sort -n\n",
    "\n",
    "\n",
    "# Count which features have no score\n",
    "echo\n",
    "echo \"Count of features that have no score\"\n",
    "awk '$6==\".\" {print $3}' $infile | sort | uniq -c\n",
    "\n",
    "\n",
    "# Find how many bad coords there are\n",
    "echo\n",
    "bad_coords=$(awk '$5 < $4' $infile | wc -l | awk '{print $1}')\n",
    "echo \"Records with bad coordinates: $bad_coords\"\n",
    "\n",
    "\n",
    "\n",
    "#_______________________________________________________________#\n",
    "#                                                               #\n",
    "#      WARNING: the following examples are more advanced!       #\n",
    "#_______________________________________________________________#\n",
    "\n",
    "# if there were records with bad coords, find the sources responsible\n",
    "if [ $bad_coords != 0 ]\n",
    "then\n",
    "    echo\n",
    "    echo \"Sources of bad coordinates:\"\n",
    "    # Instead getting one source per line, pipe into awk again to print them\n",
    "    # on one line with semicolon and space between the names\n",
    "    awk '$5 < $4 {print $2}' $infile | sort -u | awk '{sources=sources\"; \"$1} END{print substr(sources, 3)}'\n",
    "fi\n",
    "\n",
    "\n",
    "\n",
    "# Count of the features. Instead of using awk .... |sort | uniq -c\n",
    "# we will just use awk. Compare this with the above method\n",
    "# used to count the sources. Although it is a longer command, it is more efficient\n",
    "echo\n",
    "echo \"Count of each feature:\"\n",
    "awk '{counts[$3]++} END{for (feature in counts){print feature\"\\t\"counts[feature]}}' $infile | sort -k2n\n",
    "\n",
    "\n",
    "# This example is even more complicated! It uses a loop to\n",
    "# get the mean score of the genes, broken down by source.\n",
    "echo\n",
    "echo \"Getting mean scores for each source...\"\n",
    "for source in `awk '{print $2}' $infile | sort | uniq`\n",
    "do\n",
    "    awk -v s=$source '$2==s {total+=$6; count++} END{print \"Mean score for\", s\":\\t\", total/count}' $infile\n",
    "done\n",
    "\n",
    "\n",
    "# We can use awk to split the input into multiple output files.\n",
    "# Writing print \"line\" > filename will append the string \"line\"\n",
    "# to a file called filename. If a file called filename\n",
    "# does not exist already, then it will be created.\n",
    "#\n",
    "# Write a new gff for each of the sources in the original input gff file\n",
    "echo\n",
    "echo \"Writing a file per source of the original gff file $infile to files called split.*\"\n",
    "awk '{filename=\"split.\"$2\".gff\"; print $0 > filename}' $infile\n",
    "echo \" ... done!\"\n",
    "```"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
